### PR TITLE
Feat: No maturity date

### PR DESCRIPTION
**Pull request by mustermeiszer**
# Description
Upcoming assets will have no real maturity date, but rather represent the holdings of an asset, that can be hold for infinity. Think shares, etfs, etc. 

For this reason it does not make sense to have a maturity date set for those assets. 

## Changes and Descriptions
* extend `enum Maturity` with new `None` variant
* pass maturity as `Option<Seconds>` mostly around
* error out when valuation method is dcf with no maturity &rarr; leads to overflow in the `pow()` otherwise
* set maturity to `T::Time::now()` for externally priced assets when having no maturity

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by mustermeiszer** [comment]
> set maturity to `T::Time::now()` for externally priced assets when having no maturity

@lemunozm do you think this behaviour is in any means harmful for linear pricing? IIUC this just means, that the linear pricing basically would always return the given price without adjusting it, right?

File: pallets/loans/src/entities/pricing/external.rs [hunk 1]
```diff
@@ -180,7 +180,7 @@ impl<T: Config> ExternalActivePricing<T> {
 	pub fn current_price(
 		&self,
 		pool_id: T::PoolId,
-		maturity: Seconds,
+		maturity: Option<Seconds>,
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/entities/pricing/external.rs:183]: Yes, thats also why I wanted to get this in with the current one.

File: pallets/loans/src/entities/loans.rs [hunk 2]
```diff
@@ -260,9 +260,11 @@ impl<T: Config> ActiveLoan<T> {
...
 		let now = T::Time::now();
 		match trigger {
-			WriteOffTrigger::PrincipalOverdue(overdue_secs) => {
-				Ok(now >= self.maturity_date().ensure_add(*overdue_secs)?)
-			}
+			WriteOffTrigger::PrincipalOverdue(overdue_secs) => Ok(now
+				> self
+					.maturity_date()
+					.unwrap_or(T::Time::now())
+					.ensure_add(*overdue_secs)?),
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/entities/loans.rs:267]: Yes, was mostly readability focused change. But makes it harder to follow IMO.

[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs:267]: A suggestion for readability:
```rust
match self.maturity_date() {
    Some(maturity) => Ok(now > maturity.ensure_add(*overdue_secs)?),
    None => Ok(false) 
}
```

File: pallets/loans/src/entities/pricing/external.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:183]: I do not think it requires a migration, it only adds a new `None` variant to the `Maturity enum`. Note that the `Option` is never stored.

**Review by lemunozm — COMMENTED** [review]
LGTM! Very low invasive change with the implications it has!

I can not find any leak 👍🏻

File: pallets/loans/src/tests/util.rs [hunk 3]
```diff
@@ -81,6 +81,44 @@ pub fn base_internal_pricing() -> InternalPricing<Runtime> {
...
+			compounding: CompoundingSchedule::Secondly,
+		},
+		collateral: ASSET_AA,
+		pricing: Pricing::Internal(dcf_internal_pricing()),
+		restrictions: LoanRestrictions {
+			borrows: BorrowRestrictions::NotWrittenOff,
+			repayments: RepayRestrictions::None,
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/tests/util.rs:120]: Or maybe even you do not need the `dcf_internal_loan()` function, just adding that `LoanInfo {...}` in the test

File: pallets/loans/src/types/mod.rs [hunk 4]
```diff
@@ -120,6 +124,7 @@ impl Maturity {
 				date.ensure_add_assign(value)?;
 				extension.ensure_sub_assign(value)
 			}
+			Maturity::None => Err(ArithmeticError::Overflow),
```
[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/types/mod.rs:127]: I do not follow you, could you rephrase?

[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/types/mod.rs:127]: T, we could just throw a DispatchError::Other

[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/types/mod.rs:127]: Ah right! we do not have `T` there. Then I would write an `Other` explaining the error 👍🏻

File: pallets/loans/src/tests/util.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/tests/util.rs]: done

File: pallets/loans/src/types/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/types/mod.rs]: done

**Comment by lemunozm** [comment]
Thanks for tackling my NITs again!
